Key support

benmuth opened the pull request:
Up/down cursor movement is working. The solution here is to note the cursor offset from the beginning of the line, and the length of the adjacent line to be moved to. If the offset is greater than the length of the adjacent line, move to the end of the adjacent line. Otherwise move to the adjacent line at the same offset.

The code for this solution is a bit messy, mostly due to dealing with edge cases, and can probably be cleaned up. I think the general idea is solid though, and should be just as fast for long documents as short ones.

I'd be happy to hear any suggestion for improvements or see an alternate implementation.

burntcarrot commented on client/editor/editor.go:
```diff
@@ -126,18 +126,142 @@ func (e *Editor) showPositions() {
 
-	if newCursor < 0 {
-		newCursor = 0
+	// move cursor vertically
+	if y > 0 {
+		newCursor = e.calcCursorDown()
+	}
+
+	if y < 0 {
+		newCursor = e.calcCursorUp()
 	}
+
```
Could this be condensed into a if-else block?

benmuth replied:
I thought keeping them separate might be preferable because we want to do nothing if y == 0.

burntcarrot replied:
Ah, makes complete sense. Let's keep it as it is.

burntcarrot commented on client/editor/editor.go:
```diff
@@ -126,18 +126,142 @@ func (e *Editor) showPositions() {
+		pos--
+	}
+
+	cls := -1
```
I'm sorry, but what does `cls` exactly do? Does it act like a placeholder for the current line length?

burntcarrot replied:
Resolved in https://github.com/burntcarrot/rowix/pull/24#issuecomment-1367453114.

burntcarrot commented on client/editor/editor.go:
```diff
@@ -126,18 +126,142 @@ func (e *Editor) showPositions() {
+		offset = e.Cursor - cls
+	}
+
+	pls := -1
```
Does `pls` act like a placeholder for the previous line length?

burntcarrot replied:
Resolved in https://github.com/burntcarrot/rowix/pull/24#issuecomment-1367453114.

burntcarrot reviewed (commented):
I'm confused what `cls`, `pls`, `nls` do here; could you please explain their functionality in brief?

benmuth commented:
> I'm confused what `cls`, `pls`, `nls` do here; could you please explain their functionality in brief?

Sure, those stand for 'current line start' and 'previous line start'. There's also `cle` and `nle` for 'current line end' and 'next line end'. These variables hold the location of the relevant newline character in the document text. As an example:

    `f o o \n`
    `b a r \n`
    `b a z \n`

If my cursor was on the first character on the second line, `cls` would be the first `\n`, `cle` would be the second `\n`, and `nle` would be the third `\n` (if we were moving upward we would need to know `pls` as well, but here `pls` would not be set because the 'previous line start' is the beginning of the document, so there is no `\n` there).

These are used to calculate the length of the line (for instance, when moving the cursor down a line, 'next line end' - 'current line end' = length of next line) to see where the cursor should be placed. If the cursor 'offset' from the start of the line is greater than the length of the next line, then the cursor should just be placed at the end of the next line. Or, if nle-cle < offset {return nle}.

I kept those variables separate instead of just storing the line length in its own variable, because the line length is only necessary for one check (to see if the offset is greater than the length of the line to move to) while the separate variables are useful for checking edge cases like moving to the first or last line in the document.

burntcarrot commented:
Thank you for the explanation! It would be great if you could:

* add comments describing this behavior (not as inline comments, but as a large block of comment describing this in detail)
* write some unit tests for the new behavior

Writing unit tests will help in finding out certain flaws with the current implementation.

benmuth commented:
Sorry for the crazy delay! I added test cases and found bugs (of course) but got sidetracked with some other stuff before I got around to fixing it. 
It seems to be working now, and I cleaned up the code a bit and added comments as explanation.

burntcarrot commented:
Thanks for the update! Looks good to me; I'll try testing it locally once.

burntcarrot reviewed (commented):
Tried running this locally, but I still have a character remaining in the editor, and I cannot delete it:

```
a
```

I'm not sure where this is originating from. Might need to take a second look.

benmuth commented:
Right, my mistake. That was a holdover of a previous fix to handle panics that could occur with empty documents (e425b85). Should be working now.

burntcarrot commented:
@benmuth Great work on this one, thanks! Merging!
